src/main/java/seedu/address/model/group/UniqueGroupList.java

```diff
@@ -0,0 +1,150 @@
+    public boolean contains(Group toCheck) {
+        requireNonNull(toCheck);
+        return internalList.stream().anyMatch(toCheck::isSameGroup);
+    }
+
+    /**
+     * Adds a person to the list.
+     * The person must not already exist in the list.
+     */
+    public void add(Group toAdd) {
+        requireNonNull(toAdd);
+        if (contains(toAdd)) {
+            throw new DuplicatePersonException();
+        }
+        internalList.add(toAdd);
+    }
+
+    /**
+     * Replaces the group {@code target} in the list with {@code editedGroup}.
+     * {@code target} must exist in the list.
+     * The group identity of {@code editedGroup} must not be the same as another existing group in the list.
+     */
+    public void setGroup(Group target, Group editedGroup) {
+        requireAllNonNull(target, editedGroup);
+
+        int index = internalList.indexOf(target);
+        if (index == -1) {
+            throw new GroupNotFoundException();
+        }
+
+        if (!target.isSameGroup(editedGroup) && contains(editedGroup)) {
+            throw new DuplicateGroupException();
+        }
+
+        internalList.set(index, editedGroup);
+    }
+
+    /**
+     * Removes the equivalent group from the list.
+     * The group must exist in the list.
+     */
+    public void remove(Group toRemove) {
+        requireNonNull(toRemove);
+        boolean groupFound = false;
+        Group groupToDelete = null;
+        for (int i = 0; i < internalList.size(); i++) {
+            Group group = internalList.get(i);
+            if (group.isSameGroup(toRemove)) {
+                groupFound = true;
+                groupToDelete = group;
+                break;
+            }
+        }
+        if (!groupFound) {
+            throw new GroupNotFoundException();
+        }
+        internalList.remove(groupToDelete);
+    }
+
+    public void setGroups(UniqueGroupList replacement) {
+        requireNonNull(replacement);
+        internalList.setAll(replacement.internalList);
+    }
+
+    /**
+     * Replaces the contents of this list with {@code persons}.
+     * {@code persons} must not contain duplicate persons.
+     */
+    public void setGroups(List<Group> groups) {
+        requireAllNonNull(groups);
+        if (!groupsAreUnique(groups)) {
+            throw new DuplicatePersonException();
+        }
+
+        internalList.setAll(groups);
+    }
+
+    /**
+     * Returns the backing list as an unmodifiable {@code ObservableList}.
+     */
+    public ObservableList<Group> asUnmodifiableObservableList() {
+        return internalUnmodifiableList;
+    }
+
+    @Override
+    public Iterator<Group> iterator() {
+        return internalList.iterator();
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof UniqueGroupList // instanceof handles nulls
+                && internalList.equals(((UniqueGroupList) other).internalList));
```

Comment:
Same for here too

Reply:
@marcuspeh The GroupName and UniqueGroupList portion was created by @ramapriyan912001. Perhaps he can modify these parts and make a PR to update the master branch accordingly.

Reply:
@marcuspeh 
We have a preliminary check if they are the exact same object, if so, we simply return true and end there.
OR
We check that the object to compare is an instance of UniqueGroupList, and we check that the list fields contain the same group objects to check for equality.

Reply:
What I meant you can split `(other instanceof GroupName && name.equals(((GroupName) other).name))` into a variable first before using that variable in return. Something like what's shown in the screenshot.
![image](https://user-images.githubusercontent.com/54053213/135785013-9de222a1-a4a5-418d-80c1-5268aa31fc47.png)

Reply:
I see - I followed the template provided by ab3's pre-existing `UniquePersonList.java` class

```
    @Override
    public boolean equals(Object other) {
        return other == this // short circuit if same object
                || (other instanceof UniquePersonList // instanceof handles nulls
                        && internalList.equals(((UniquePersonList) other).internalList));
    }
```

Do you feel we should stick to the template? Or try to simplify?

Reply:
I think for now we stick on first. We can do a code quality check after we are done with v1.2 and change all of theses